### ISSUE TITLE
Error Channel Handling Incorrect

### ISSUE THREAD

**Issue by garyrussell** [spring-cloud/spring-cloud-stream#2021]
Given

```java
    @ServiceActivator(inputChannel = "trades.errors")
    public void error(Message<?> message) {
        logger.error("Received on trades.errors {}", message);

        doError(message);
    }
```

Spring integration will create a `DirectChannel`. The `AbstractMessageChannelBinder` ensures that a user-defined error channel is subscribable but not that it is pub/sub, causing round robin distribution (between the service and the bridge to the global error handler).

Work around is to explicitly declare it as a pubsub.

I see several options:

1. Detect and expect a pub/sub channel, throwing an exception.
2. If the channel is not pub/sub, don't bridge to the global channel (and log a warn?).
3. ...?

I think I lean towards 2.

**Comment by olegz** [comment]
I am thinking 2 as well

**Comment by olegz** [comment]
Actually I still can't reproduce it. 
From the code snippet above it appears to be producer's output channel since consumers would need to include group name.
Going through `doBindProducer` code in `AbstractMessageChannelBinder` the `registerErrorInfrastructure` return an instance of `PublisheSubscribeChannel` for the binding specific error channel.

Also, while it may not be relevant to this issue consumer side binding error channel is an instance of `BinderErrorChannel` which is also `PublisheSubscribeChannel`.

In other words both channels are PubSub, so perhaps more info

**Comment by garyrussell** [comment]
```java
@SpringBootApplication
public class Stgh2021Application {

	public static void main(String[] args) {
		SpringApplication.run(Stgh2021Application.class, args);
	}

	@Autowired
	StreamBridge bridge;

	@Bean
	public ApplicationRunner runner(ApplicationContext ctx) {
		return args -> {
			bridge.send("output", "foo");
			bridge.send("output", "bar");
			System.out.println(ctx.getBean("foo.errors").getClass());
		};
	}

	@ServiceActivator(inputChannel = "foo.errors")
	void error(Message<?> msg) {
		System.out.println(msg);
	}

}
```
```properties
spring.cloud.stream.bindings.output.destination=foo
spring.cloud.stream.bindings.output.producer.error-channel-enabled=true

spring.rabbitmq.publisher-returns=true
```
First failure goes to the SA, second one goes to the logging handler:

```
class org.springframework.integration.channel.DirectChannel
ErrorMessage [payload=org.springframework.integration.amqp.support.ReturnedAmqpMessageException, failedMessage=GenericMessage [payload=byte[3], headers={id=b7c7ce73-7af2-c8a3-2930-3a80abacfd65, timestamp=1602519480243}] [amqpMessage=(Body:'bar' MessageProperties [headers={spring_returned_message_correlation=808ed74c-d365-e368-6b00-722ea071d88c, accept=application/json}, timestamp=Mon Oct 12 12:18:00 EDT 2020, messageId=808ed74c-d365-e368-6b00-722ea071d88c, contentType=application/json, contentLength=0, receivedDeliveryMode=PERSISTENT, priority=0, deliveryTag=0]), replyCode=312, replyText=NO_ROUTE, exchange=foo, routingKey=foo], headers={id=c1cc2faf-5861-d21f-040a-b423207610d3, timestamp=1602519480243}]
2020-10-12 12:18:00.251 ERROR 32065 --- [tory.publisher1] o.s.integration.handler.LoggingHandler   : org.springframework.integration.amqp.support.ReturnedAmqpMessageException, failedMessage=GenericMessage [payload=byte[3], headers={id=b386c2ef-b16a-ab5b-28ba-2fdc2a9b8440, timestamp=1602519480243}] [amqpMessage=(Body:'foo' MessageProperties [headers={spring_returned_message_correlation=9df1bcfe-aab0-a676-c8a2-124f775ce630, accept=application/json}, timestamp=Mon Oct 12 12:18:00 EDT 2020, messageId=9df1bcfe-aab0-a676-c8a2-124f775ce630, contentType=application/json, contentLength=0, receivedDeliveryMode=PERSISTENT, priority=0, deliveryTag=0]), replyCode=312, replyText=NO_ROUTE, exchange=foo, routingKey=foo]
	at org.springframework.integration.amqp.outbound.AbstractAmqpOutboundEndpoint.buildReturnedMessage(AbstractAmqpOutboundEndpoint.java:620)
	at org.springframework.integration.amqp.outbound.AmqpOutboundEndpoint.returnedMessage(AmqpOutboundEndpoint.java:275)
	at org.springframework.amqp.rabbit.core.RabbitTemplate.handleReturn(RabbitTemplate.java:2510)
	at org.springframework.amqp.rabbit.connection.PublisherCallbackChannelImpl.lambda$handleReturn$2(PublisherCallbackChannelImpl.java:1100)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630)
	at java.base/java.lang.Thread.run(Thread.java:832)
```

**Comment by olegz** [comment]
It's a bit tricky since the channel is created by SI in `AbstractMethodAnnotationPostProcessor:354`
```
catch (DestinationResolutionException e) {
	if (e.getCause() instanceof NoSuchBeanDefinitionException) {
		inputChannel = new DirectChannel();
. . . .
```
I'll see if I can bridge it
